src/System.Runtime.Numerics/src/System/Numerics/Complex.cs

```diff
@@ -295,7 +313,90 @@ public static Complex Exp(Complex value)
+                //   x = \sqrt{\frac{\sqrt{a^2 + b^2} + a}{2}}  y = \pm \sqrt{\frac{\sqrt{a^2 + b^2} - a}{2}}
+                // There is just one complication: depending on the sign on a, either x or y suffers from
+                // cancelation when |b| << |a|. We can get aroud this by noting that our formulas imply
+                // x^2 y^2 = b^2 / 4, so |x| |y| = |b| / 2. So after computing the one that doesn't suffer
+                // from cancelation, we can compute the other with just a division. This is basically just
+                // the right way to evaluate the quadratic formula without cancelation.
+
+                // All this reduces our total cost to two sqrts and a few flops, and it respects the desired
+                // symmetries. Much better than atan + cos + sin!
+
+                // The signs are a matter of choice of branch cut, which is traditionally taken so x > 0 and sign(y) = sign(b).
+      
+                // If the components are too large, Hypot will overflow, even though the subsequent sqrt would
+                // make the result representable. To avoid this, we re-scale (by exact powers of 2 for accuracy)
+                // when we encounter very large components to avoid intermediate infinities.
+                bool rescale = false;
+                if ((Math.Abs(value._real) >= s_sqrtRescaleThreshold) || (Math.Abs(value._imaginary) >= s_sqrtRescaleThreshold))
+                {
+                    if (Double.IsInfinity(value._imaginary) && !Double.IsNaN(value._real))
+                    {
+                        // We need to handle infinite imaginary parts specially because otherwise
+                        // our formulas below produce inf/inf = NaN. The NaN test is necessary
+                        // so that we return NaN rather than (+inf,inf) for (NaN,inf).
+                        return (new Complex(Double.PositiveInfinity, value._imaginary));
+                    }
+                    else
+                    {
+                        value._real *= 0.25;
+                        value._imaginary *= 0.25;
+                        rescale = true;
+                    }
+                }
+ 
+                // This is the core of the algorithm. Everything else is special case handling.
+                double x, y;
+                if (value._real >= 0.0)
+                {
+                    x = Math.Sqrt((Hypot(value._real, value._imaginary) + value._real) * 0.5);
+                    y = value._imaginary / (2.0 * x);
+                }
+                else
+                {
+                    y = Math.Sqrt((Hypot(value._real, value._imaginary) - value._real) * 0.5);
+                    if (value._imaginary < 0.0) y = -y;
+                    x = value._imaginary / (2.0 * y);
+                }
+
+                if (rescale)
+                {
+                    x *= 2.0;
+                    y *= 2.0;
+                }
```

Comment:
@dcwuser Good catch. The fix is simple:

    t = 2.0 * Math.Sqrt(absReal * (0.125 * (1.0 + Math.Sqrt(1.0 + ratio * ratio))));

and similarly for the else clause. The last factor is always less than 1, so you won't get overflow. I prefer this 'local' solution to wrapping the whole calculation in a scale/unscale duo.

Now, there is still a problem here: for about 10<sup>-35</sup>% of inputs, this will underflow to zero: when both arguments are denormals less than `4*double.Epsilon`. If and only if underflow occurs, `t` is zero, so you can easily test for this exceptional case before the final if-statement, and use the original expression for t, which doesn't have issues for these inputs:

    if (t == 0)
        t = Math.Sqrt(0.5 * (absReal + Hypot(absReal, absImaginary)));